### PR TITLE
Update disallow-helm-tiller and disallow-latest-tag to include all container types in a pod

### LINKED ISSUE #951: [Bug] update sample policies to include all container types in a pod
### Kyverno Version

1.12

### Kubernetes Version

1.29

### Kubernetes Platform

Minikube

### Description

A end user brought this up at KubeCon EU 2024. A number of sample policies target containers in a pod, but do not handle initContainers and ephemeralContainers.  

Note that this is not an issue for any Pod Security Standard related policies, but applies to best practices and other security samples.

However, this may be something end users are oblivious of and hence end up using the sample policy which leaves a potential gap.  

This can be handled easily as done in the following policy:

https://kyverno.io/policies/psp-migration/restrict-adding-capabilities/restrict-adding-capabilities/

```yaml
      validate:
        message: >-
          Any capabilities added other than NET_BIND_SERVICE or CAP_CHOWN are disallowed.          
        foreach:
          - list: request.object.spec.[ephemeralContainers, initContainers, containers][]
            deny:
              conditions:
                all:
                - key: "{{ element.securityContext.capabilities.add[] || '' }}"
                  operator: AnyNotIn
                  value:
                  - NET_BIND_SERVICE
                  - CAP_CHOWN
                  - ''
```

Here is a list of some of these policies:

- [ ] handle all container types: https://kyverno.io/policies/psp-migration/add-apparmor/add-apparmor/
- [ ] handle all container types: https://kyverno.io/policies/psp-migration/add-capabilities/add-capabilities/
- [ ] handle all container types: https://kyverno.io/policies/other/add-default-resources/add-default-resources/
- [ ] handle all container types: https://kyverno.io/policies/other/allowed-base-images/allowed-base-images/

### Steps to reproduce

1. Check samples like: https://kyverno.io/policies/psp-migration/add-apparmor/add-apparmor/

### Expected behavior

1. Update all security and best practices related policies to handle all pod types.
2. Update test cases to handle initContainers

### Screenshots

_No response_

### Kyverno logs

_No response_

### Slack discussion

_No response_

### Troubleshooting

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have searched other issues in this repository and mine is not recorded.

### PR DESCRIPTION
**Pull request by dolisss**
## Related Issue(s)

Partially addresses #951

## Description

Updating disallow-helm-tiller and disallow-latest-tag policies to handle all pod types and  test cases to handle initContainers. 

## Checklist

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

### CODE REVIEW

**Comment by dolisss** [comment]
Fixes

**Review by chipzoller — CHANGES_REQUESTED** [review]
Tiller (Helm v2) is so old technology I don't think it's worth making this change. Would suggest focusing on more useful policies in the present.

**Comment by dolisss** [comment]
I am updating all the best practice policies. Can you approve this one while I add the logic in other policies too?

**Comment by dolisss** [comment]
@chipzoller  Extended the logic to the "disallow latest tag" policy along with changes needed for the chainsaw test cases.

**Review by chipzoller — CHANGES_REQUESTED** [review]
Missing extended tests for disallow-helm-tiller.

**Comment by dolisss** [comment]
@chipzoller  thought the Helm Tiller policy was outdated and that test cases might not be necessary. However, I’ve updated them now. Please review the changes

**Comment by chipzoller** [comment]
How have you tested the changes you are making in this PR?

**Comment by dolisss** [comment]
I have tested this locally as well as in Kyverno Playground.Do you see any issue? Let me know..

File: best-practices/disallow-latest-tag/disallow-latest-tag.yaml [hunk 1]
```diff
@@ -25,10 +25,16 @@ spec:
           - Pod
     validate:
       message: "An image tag is required."
-      pattern:
-        spec:
-          containers:
-          - image: "*:*"
+      foreach:
+        - list: "request.object.spec.initContainers"
```
Comment (chipzoller) [inline best-practices/disallow-latest-tag/disallow-latest-tag.yaml:29]: No containers[] any longer.

File: best-practices/disallow-latest-tag/disallow-latest-tag.yaml [hunk 2]
```diff
@@ -37,7 +43,13 @@ spec:
           - Pod
     validate:
       message: "Using a mutable image tag e.g. 'latest' is not allowed."
-      pattern:
-        spec:
-          containers:
-          - image: "!*:latest"
\ No newline at end of file
+      foreach:
+        - list: "request.object.spec.initContainers"
```
Comment (chipzoller) [inline best-practices/disallow-latest-tag/disallow-latest-tag.yaml:47]: No containers[] any longer.

File: best-practices/disallow-helm-tiller/disallow-helm-tiller.yaml [hunk 3]
```diff
@@ -26,8 +26,13 @@ spec:
               - Pod
       validate:
         message: "Helm Tiller is not allowed"
-        pattern:
-          spec:
-            containers:
-              - name: "*"
-                image: "!*tiller*"
+        foreach:
+        - list: "request.object.spec.initContainers"
```
Comment (chipzoller) [inline best-practices/disallow-helm-tiller/disallow-helm-tiller.yaml:30]: No containers[] any longer.

File: best-practices/disallow-helm-tiller/artifacthub-pkg.yml [hunk 4]
```diff
@@ -18,4 +18,4 @@ readme: |
 annotations:
   kyverno/category: "Sample"
   kyverno/subject: "Pod"
-digest: 6de64a4a8d611c250dc0190b28b6c757db531063161531e4f68202c0fbda5be4
+digest: 31126cee424a796fe9a4078d879a2d650fa6a3efc267d714a90a54604d91a9de
```
Comment (chipzoller) [inline best-practices/disallow-helm-tiller/artifacthub-pkg.yml:21]: Recalculate.

File: best-practices/disallow-latest-tag/artifacthub-pkg.yml [hunk 5]
```diff
@@ -18,4 +18,4 @@ readme: |
 annotations:
   kyverno/category: "Best Practices"
   kyverno/subject: "Pod"
-digest: 3d19e0d8f2637eca9ad1d700f4fbf556aaa31221ff6c40698b9aadda1f41adb4
+digest: 3e299e0f018ff57ee78cfff4119df58d72e11669908eff9dfe8294f3be7a6e60
```
Comment (chipzoller) [inline best-practices/disallow-latest-tag/artifacthub-pkg.yml:21]: Recalculate.

**Review by chipzoller — CHANGES_REQUESTED** [review]

**Review by chipzoller — CHANGES_REQUESTED** [review]
Please fix failing tests. Also complete the PR template (checklist).

**Comment by dolisss** [comment]
@chipzoller I tried updating the digest using my Mac, but the check failed. I reverted to the old digest, but it still didn't work. How can I generate the digest? I am using the following command.

```
 shasum -a 256 disallow-latest-tag.yaml 
2760272e57d9988ba447f62d23bba382092d00a5e14dbf00555e4170ea90593a 
```

**Review by chipzoller — CHANGES_REQUESTED** [review]
Digests are expected to be generated in Linux; you can see the digest expected in the Policy Test step and use that if you're unable to generate one on your own system.

Tests still need to be fixed here.

**Comment by dolisss** [comment]
@chipzoller Thanks! it worked with Linux. What checks are you referring to? I don't see any on my page.

**Comment by chipzoller** [comment]
Please see failing CI checks.

**Comment by dolisss** [comment]
@chipzoller Please review.

**Review by chipzoller — APPROVED** [review]

**Pull request merged**